File: frontend/src/types/habit.ts
export interface HabitLog {
  id: number
  habit_id: number
  date: string
  completed_at: string
}

export interface Habit {
  id: number
  name: string
  description: string | null
  color: string
  is_active: boolean
  created_at: string
  logs: HabitLog[]
}

export interface HabitCreate {
  name: string
  description?: string | null
  color?: string
}

export interface HabitUpdate {
  name?: string
  description?: string | null
  color?: string
  is_active?: boolean
}

File: backend/app/schemas/habit.py
from datetime import date, datetime

from pydantic import BaseModel, Field

COLOR_PATTERN = r"^#[0-9A-Fa-f]{6}$"


class HabitCreate(BaseModel):
    name: str = Field(max_length=255)
    description: str | None = None
    color: str = Field(default="#10B981", pattern=COLOR_PATTERN)


class HabitUpdate(BaseModel):
    name: str | None = Field(default=None, max_length=255)
    description: str | None = None
    color: str | None = Field(default=None, pattern=COLOR_PATTERN)
    is_active: bool | None = None


class HabitLogToggle(BaseModel):
    date: date


class HabitLogResponse(BaseModel):
    id: int
    habit_id: int
    date: date
    completed_at: datetime

    model_config = {"from_attributes": True}


class HabitResponse(BaseModel):
    id: int
    name: str
    description: str | None
    color: str
    is_active: bool
    created_at: datetime
    logs: list[HabitLogResponse] = []

    model_config = {"from_attributes": True}
